Add P3 to EAM

AaronDonahue opened the pull request:
This commit introduces two new files,

micro_p3_interface.F90
micro_p3_utils.F90

and makes fundamental changes to the file,

micro_p3.F90

The interface code sets up an interface between the E3SM infrastructure
and the main P3 subroutines.  Similar in structure to the interface
for MG, the interface code initializes P3, initializes the PBUF variables
associated with P3, converts physics state variables to work with P3,
and conducts any post-processing of P3 output to be compatible with E3SM.

The utils code includes utility functions necessary to make P3 work,
including the loading of all constants used in P3.

There was an inconsistentcy between how E3SM defines the ratio
between temperature and potential temperature (exner) which has been
fixed.

The main micro_p3 routines have been adjusted to be consistent with
E3SM and to recieve the appropriate input/output variables needed to
interface with E3SM.

The microp_driver has been changed to call P3 routines when P3 microphysics
is chosen as the microphysics scheme.

Lastly, there were some minor configuration changes needed to get this
version of E3SM running properly on LC machines.

pressel commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -673,10 +552,23 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th_old,th,qv_old,qv,dt,qitot,qirim,nitot,birim,ss
+    ! AaronDonahue added exner term to replace all instances of th(i,k)/t(i,k), since th(i,k) is updated but t(i,k) is not, and this was
+    ! causing energy conservation errors.
+    inv_exner = 1./exner        !inverse of Exner expression, used when converting potential temp to temp
+    tmparr1 = inv_exner !(pres*1.e-5)**(rd*inv_cp)
```
What is the purpose of tmparr1 here? Could you just use inv_exner to improve readability?

AaronDonahue replied:
I agree that it would just be simpler to use inv_exner on the next line instead of having a tmparr1.  I left it in because I didn't know how many changes we wanted to make from the original P3.  I'm happy to make the change now, or we will probably have a code clean-up stage where we can address things like this.

pressel replied:
Ok, that sounds fine to me. Let's push this off to code clean-up.

PeterCaldwell replied:
might as well make this small change now...

AaronDonahue replied:
Sounds good

pressel commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -1665,9 +1559,9 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th_old,th,qv_old,qv,dt,qitot,qirim,nitot,birim,ss
+          th(i,k) = th(i,k) + exner(i,k)*((qidep-qisub+qinuc)*     &
                xxls(i,k)*inv_cp +(qrcol+qccol+qchetc+   &
-               qcheti+qrhetc+qrheti-qimlt)*       &
+               qcheti+qrhetc+qrheti-qimlt)*       &  !AaronDonahue - Is qrmul missing here?
```
My knowledge of rime-splintering is small but this does seem problematic if it is the case that splintering does involve a phase change (which I think it should). Could the release of the latent heat fusion associated with the phase be accounted for in one of the other source terms, if so we should check carefully that we are not double accounting for the water source/sinks associate with splintering.

AaronDonahue replied:
Yes, hence my comment.  Having qrmul not be listed here made me uncomfortable. It is a source/sink for ice/rain respectively so you would think it should be included in the temperature tendency.  That being said, qrmul is never defined in the actual code and thus is 0, so it doesn't impact anything.  I would feel better including qrmul here but I wanted to get input from someone with more familiarity with microphysics to make sure I'm not missing something.  @PeterCaldwell , do you have an opinion?

PeterCaldwell replied:
I agree that qrmul should affect the temperature. I'm uncomfortable with it being modified (on line 1590) without ever having been set explicitly to zero, and also with how qcmul is declared but never used at all. Presumably qcmul = -qrmul? I suggest we delete all mention of either of these variables. If we decide to add them later, we can start from a clean slate and do it right.

AaronDonahue replied:
Sounds good, I can do that.  Would it be worth determining what (if) other tendency terms are defined and never set/used and remove them as well?  Perhaps including a comment about which tendencies have been given this treatment.

PeterCaldwell replied:
Yeah, searching for each defined variable and ensuring it is defined somewhere would probably not take too long...

AaronDonahue replied:
not at all.  I'll go ahead and do that.

pressel commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -1588,7 +1482,7 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th_old,th,qv_old,qv,dt,qitot,qirim,nitot,birim,ss
              qrheti = qrheti*ratio
              qrmul  = qrmul*ratio
-             !qrhetc = qrhetc*ratio
+             !qrhetc = qrhetc*ratio   ! AaronDonahue - Why is this commented out? We don't have qrhetc?
```
Do we have an answer to this question yet?

AaronDonahue replied:
No.  I just flagged this because I thought it was strange to have it written and then commented out.  This is another tendency that isn't defined and so is 0 anyway, so commented or not it doesn't impact the solution.  Still, if in the future we do include qrhetc this seems like a potential bug.  Should we fix?  @PeterCaldwell , and thoughts?

PeterCaldwell replied:
Yeah, this is another case of a variable that is never assigned. I think we should delete it as well, ensuring that we'll do a complete job if we decide to include it again later.

AaronDonahue replied:
Check, I've deleted this in the new commit.

pressel commented on components/cam/src/physics/cam/micro_p3_interface.F90:
```diff
@@ -0,0 +1,1555 @@
+    ! HANDLE AEROSOL ACTIVATION
+    !==============
+    !saving this for later... use prescribed Nd for now.
+    log_predictNc = .false.
```
Do you want log_predictNc hardwired to false in this PR? I'm ok with leaving this way for now as I will make it a namelist option with my upcoming PR.

AaronDonahue replied:
Yes, I think that is a good plan forward since this PR represents the initial port of P3 and later PR's will incorporate improvements.

ambrad commented:
Am I correct that this PR is not intended to bring in the interpolator fix re: find_lookupTable_indices_3? I see lines of the sort
```
            dum1 = revap_table(dumii,dumjj)+(rdumii-real(dumii))*inv_dum3*                   &
                  (revap_table(dumii+1,dumjj)-revap_table(dumii,dumjj))
```
which still have the spurious inv_dum3 factor.

AaronDonahue commented:
@ambrad , I am unaware of the interpolator fix.  Could you give more details?  I know that in the next version of P3 for WRF they have changed the lookup tables, is this what you are referring too?  If so, then I would like to put in the fixes for using v4 lookup tables in a later PR.  But if the fix you are referring to is more of a known bugfix I think we should put it either now or issue a PR soon that fixes the issue.

PeterCaldwell commented:
Argh, sorry Aaron - I thought you were aware of the bug Andrew found in the table lookup (which is the reason why it is fixed in the most recent WRF version ;-) ). I'm somewhat inclined to say let's fix the table lookup bug in its own PR, but would be happy fixing it here too.

AaronDonahue commented:
Oh yes, I remember the email correspondence with Hugh.  I'm inclined to make the fix as it's own PR, but am happy to make the changes here if that is the consensus.

jgfouca commented:
We also need to deal with the build error in the CI testing:

```
/home/travis/build/E3SM-Project/scream/components/cam/src/physics/cam/micro_p3.F90:40:7:
    use spmd_utils,       only: masterproc
       1
Fatal Error: Can't open module file ‘spmd_utils.mod’ for reading at (1): No such file or directory
```

AaronDonahue commented:
The Travis-CI error is a strange one.  It's the same line that is included in micro_mg_cam.F90 which I presume is compiled at roughly the same time as the micro_p3.F90 code.  Does anyone have any thoughts on why loading the SPMD_UTILS module in micro_p3 would be a problem but not in MG?

AaronDonahue commented:
So as a list of changes to make to get this PR ready would be:
1) Find all microphysics tendencies that are defined but not used and remove them (e.g. qrmul)
2) Fix redundant code with the tmparr1 variable
3) Figure out what is causing the Travis-CI error with spmd_utils

Finally, the lookup table bugfix will be added as a bugfix in a subsequent PR?

AaronDonahue commented:
@PeterCaldwell , In doing task #1 from above I found another strange variable, qwgrth defined as the wet growth rate.  It is defined and actually gets a value but isn't used anywhere.  There is a section of commented out code that implies it could be used at some point, but not sure...  Could you take a look?

```          !----
          ! !      ! modify due to latent heating from riming rate
          ! !      !   - currently this is done by simple linear interpolation
          ! !      !     between conditions for dry and wet growth --> in wet growth it is assumed
          ! !      !     that particle surface temperature is at 0 C and saturation vapor pressure
          ! !      !     is that with respect to liquid. This simple treatment could be improved in the future.
          ! !        if (qwgrth.ge.1.e-20) then
          ! !           dum = (qccol+qrcol)/qwgrth
          ! !        else
          ! !           dum = 0.
          ! !        endif
          ! !        dumqvi = qvi(i,k) + dum*(qvs(i,k)-qvi(i,k))
          ! !        dumqvi = min(qvs(i,k),dumqvi)
          !====```

AaronDonahue commented:
@jgfouca , I'm at a loss as to why the build is failing for the `spmd_utils` mod.  I did a search and that same line is all over the code, using `spmd_utils` module to get the `masterproc` variable.  I've compiled the code on LLNL machines both in debug and not in debug mode with no issues.  The commit I just submitted takes the line verbatim from the MG microphysics equivalent file.

Any thoughts on how to fix this?

ambrad commented:
@jgfouca is it possible that updating `components/scream/src/physics/p3/CMakeLists.txt` to include more F90 files in the `${SCREAM_BASE_DIR}/../cam/src/physics/cam` directory will fix things?

AaronDonahue commented:
@ambrad , I updated my comment above.  It looks like it doesn't have to do with the module itself.  Instead it just doesn't like the first module call.

AaronDonahue commented:
Is the build error sensitive to case?  I noticed that I have `MODULE MICRO_P3` whereas most of the modules would have `module micro_p3`.  I would think it isn't since Fortran isn't case sensitive, but maybe...

ambrad commented:
@jgfouca @AaronDonahue this is a result of needing to resolve dependencies in the scream testing framework, such as spmd_mod, cam_logfile, shr_kind_mod, and their dependencies. Travis CI isn't the issue; just doing the basic scream build on any platform will fail.

Jim, I recommend you pull this PR locally and try to build. You'll get a build failure. Then work on an approach to resolve the dependencies. It may not be trivial b/c of subdependencies.

AaronDonahue commented:
@PeterCaldwell , I went ahead and added the intrpolate bugfix from P3 v3.1.8 that fixes the bug that @ambrad found, rather than leave it for a later PR.

Aside from the subdependencies issue discussed above I think I've addressed all the issues brought up.  Let me know if anything else needs to be addressed.

jgfouca commented:
@AaronDonahue , sorry for slow response. I'll work on the build issue.

jgfouca commented on components/scream/src/physics/p3/p3_ic_cases.cpp:
```diff
@@ -96,6 +96,8 @@ FortranData::Ptr make_mixed () {
     d.dzq(0,k) = Constants<double>::RD*T(k)/(g*d.pres(0,k))*dpres;
   }
 
+  // TODO - need data for pdel, exner, ast
```
@AaronDonahue , need more initial condition setup here.

jgfouca replied:
@AaronDonahue , this should probably be resolved before you merge.

AaronDonahue replied:
absolutely, I pushed a new commit just now, @jgfouca , can you check to make sure that is the correct syntax?

jgfouca replied:
There was a minor syntax problem (** is not support in C++, use std::pow instead). And I removed the TODO since it's done now.

jgfouca commented on components/scream/src/physics/p3/CMakeLists.txt:
```diff
@@ -1,10 +1,60 @@
+
+# end share/util
+
+add_definitions(-DPCOLS=1 -DPSUBCOLS=1 -DPLEV=1)
```
These definitions probably won't work. Not sure where I should get those values.

AaronDonahue commented:
As per @ambrad comment on Slack.  Is there a way we can just isolate the version that SCREAM is hooking into from the changes we are making to the Fortran version of P3?

In other words.  We will eventually be porting P3 to the C++ scream version, and when we do that we will also want to make sure the baseline tests are hooked up to that version of P3, but in the meantime we can have a hard copy of a frozen version of the micro_p3.f90 code that the scream build points too that doesn't cause any build errors (i.e. doesn't have these dependency issues).  When we have a finalized version of P3 that we want to port to scream we will a) be porting into C++ which may make it easier than having any Fortran wrappers and b) can make sure during the port that if we have any dependencies that those are brought over too in the appropriate way.

I suggest this to make the development of P3 in Fortran quicker.  We will be able to work and build and test in Fortran without worrying about the impact it has on the scream build.

There is the potential for confusion in the future when we do get around to porting P3 to SCREAM, but if we all agree that the "frozen version of micro_p3 hook" is just a temporary fix for this problem that needs to be fixed later then we should go into the porting of P3 to C++ with eyes wide open and hopefully not be tripped up.

@ambrad and @jgfouca , since you both know more about the scream build side of things does this sound like a viable plan?  Would it fix the issue for the time being?

AaronDonahue commented:
@jgfouca , @PeterCaldwell , in writing up the SCREAM-AD notes from yesterday it occurred to me that I could also change this PR such that the two codes `micro_p3.F90` and `micro_p3_utils.F90` have no dependencies and have all information explicitly passed into them.  Which is probably what the eventual C++ port will look like.  This should similarly solve the build problem I believe.

jgfouca commented:
@AaronDonahue , I've got it building on my machine. I just need to figure out why mpi fortran isn't working on the travis machine.

jgfouca commented:
OK, travis seems to be happy now.

rljacob commented:
@AaronDonahue can you change the title of this PR?  It should be "verb noun".  So "Add P3 to EAM" or something like that.   And it should not spill over into the body.

AaronDonahue commented:
@rljacob , check, I've made the change.

AaronDonahue commented:
Is this PR ready to merge now?  I want to make sure before I push the shiny green button.

ambrad commented:
Kyle and Peter should approve before you merge. I'm adding myself b/c I'm chasing down some FPEs and would like to at least understand them before merging.
